core/templates/components/button-directives/create-activity-button.component.spec.ts

```diff
@@ -0,0 +1,290 @@
+    TestBed.configureTestingModule({
+      declarations: [
+        CreateActivityButtonComponent,
+        TranslatePipe
+      ],
+      providers: [
+        UserService,
+        {
+          provide: UrlService,
+          useClass: MockUrlService
+        },
+        {
+          provide: WindowRef,
+          useValue: windowRef
+        },
+        ExplorationCreationService
+      ]
+    }).compileComponents();
+  }));
+
+  beforeEach(() => {
+    fixture = TestBed.createComponent(CreateActivityButtonComponent);
+    component = fixture.componentInstance;
+    userService = TestBed.inject(UserService);
+    urlService = TestBed.inject(UrlService);
+    explorationCreationService = TestBed.inject(ExplorationCreationService);
+    ngbModal = TestBed.inject(NgbModal);
+    fixture.detectChanges();
+  });
+
+  it('should create', () => {
+    expect(component).toBeDefined();
+  });
+
+  it('should initialize and begin creation process if user can' +
+    ' create collections', fakeAsync(() => {
+    const userServiceSpy = spyOn(userService, 'getUserInfoAsync')
+      .and.returnValue(Promise.resolve(userInfoCanCreateCollection));
+    const urlParamsSpy = spyOn(urlService, 'getUrlParams').and.returnValue({
+      mode: 'create'
+    });
+    spyOn(component, 'initCreationProcess');
+
+    expect(component.canCreateCollections).toBe(false);
+    expect(component.userIsLoggedIn).toBe(false);
+
+    component.ngOnInit();
+    tick();
+    fixture.detectChanges();
+
+    expect(userServiceSpy).toHaveBeenCalled();
+    expect(urlParamsSpy).toHaveBeenCalled();
+    expect(component.initCreationProcess).toHaveBeenCalled();
+    expect(component.canCreateCollections).toBe(true);
+    expect(component.userIsLoggedIn).toBe(true);
+  }));
+
+  it('should initialize and create new exploration if user cannot' +
+    ' create collections', fakeAsync(() => {
+    const userServiceSpy = spyOn(userService, 'getUserInfoAsync')
+      .and.returnValue(Promise.resolve(userInfoCanNotCreateCollection));
+    const urlParamsSpy = spyOn(urlService, 'getUrlParams').and.returnValue({
+      mode: 'create'
+    });
+    const explorationCreationServiceSpy = spyOn(
+      explorationCreationService, 'createNewExploration');
+
+    expect(component.canCreateCollections).toBe(false);
+    expect(component.userIsLoggedIn).toBe(false);
+
+    component.ngOnInit();
+    tick();
+    fixture.detectChanges();
+
+    expect(userServiceSpy).toHaveBeenCalled();
+    expect(urlParamsSpy).toHaveBeenCalled();
+    expect(explorationCreationServiceSpy).toHaveBeenCalled();
+    expect(component.canCreateCollections).toBe(false);
+    expect(component.userIsLoggedIn).toBe(true);
+  }));
+
+  it('should stop the creation process if another' +
+    ' creation is in progress', () => {
+    component.creationInProgress = true;
+    expect(component.initCreationProcess()).toBe();
+  });
+
+  it('should create new exploration if user cannot create collections', () => {
```

Comment:
How is this different from https://github.com/oppia/oppia/pull/12936/files#diff-c6bbc8b315fa5210ba8944e75fc8abb35cca1ca8b84b8e074ab017d687aeffccR173 ?

Reply:
There is a problem with the code. When `canCreateCollections` is `true` [here](https://github.com/oppia/oppia/blob/f753ff5929ca70a32b2fc22d5f9edb0dec8d8ccc/core/templates/components/button-directives/create-activity-button.component.ts#L97), then `initCreationProcess()` is called, but [this](https://github.com/oppia/oppia/blob/f753ff5929ca70a32b2fc22d5f9edb0dec8d8ccc/core/templates/components/button-directives/create-activity-button.component.ts#L68) line is executed only if `canCreateCollections` is false.

So, I had to check `initCreationProcess` separately, with `canCreateCollections = false`. Also, changed the name of the other tests to denote that it begins `activity` ceation.

Reply:
Why not fix the code? That's kind of the point of doing the tests -- if the tests uncover a bug, then fix the bug, rather than leave the bug there and try to test a pathway that can't occur in practice.

Reply:
I may have misjudged the code, this pathway is possible, when `initCreationProcess` is not called from `ngOnInt` but by clicking on the create button.

Regarding the comment above, I tried to completely remove the functionality to create collections, but then it was going to affect a lot of e2e tests as well. So, after discussion with @kevintab95 we decided to do that in a separate PR.

Reply:
OK, but then I go back to the earlier question ... how is this different from the already-existing similar test above?

Reply:
- This test checks the condition when - user is on the creator dashboard page & cannot create collections.
- The test above checks the condition when - user is redirected to creator dashboard after they log in & cannot create collections.

Modified the test description to state the same.